### DISCUSSION TITLE
Use of multiple exogenous time series with multiple endogenous time series

### DISCUSSION THREAD

**Discussion by yarnabrina** [JoaquinAmatRodrigo/skforecast#546] [Q&A] [answered]
Hi, I was evaluating `skforecast` with `XGBRegressor`.

1. The data I'm using is a hierarchical/panel data (with multiple levels, but hierarchy is not of immediate concern).

2. For each level of hierarchy, we have observations on multivariate target. To start with, we can have different models for each one of them.

3. There are few exogenous variables as well. Few are series independent and remains same for all (say date time based features), but rest are different for each series (and for now let's assume that those are known for future).

Is this a situation currently supported by `skforecast`?

I suppose (1) is handled by `ForecasterAutoregMultiSeries`, and (2) is handled by `ForecasterAutoregMultiVariate`, but it was not clear to me how to combine these.

But I'm completely baffled on (3), as the documentation says the following, and it seems to be unsupported (but I'm not sure).

> Exogenous variable/s included as predictor/s. Must have the same number of observations as series and their indexes must be aligned so that series[i] is regressed on exog[i].

I'll appreciate if someone (@JoaquinAmatRodrigo ?) can suggest ways to solve this. I'm completely new with `skforecast`, so may not be using same terms. If any clarification is required for any of the queries, let me know and I'll try to demonstrate with `sktime` or similar package to show what I meant.

**Comment by JavierEscobarOrtiz** [comment]
Hello @yarnabrina 

Thanks for opening the discussion. If I understand you correctly, your problem is that you want to add a different exogenous variable per series. 

Regarding the `ForecasterAutoregMultiVariate`, since adding more series to the training matrix causes it to grow horizontally (adding more columns), the exog is the same for all series.

![forecaster_multivariate_train_matrix_diagram](https://github.com/JoaquinAmatRodrigo/skforecast/assets/95030312/dbf1231b-178f-4939-9695-2b754e492d21)

On the other hand, the `ForecasterAutoregMultiSeries` grows the matrix vertically, i.e. the `lag 1` column contains all the series used during training. In this case, it is possible to have different exogenous values for each series.

Imagine that in the following diagram, the values in the red boxes are changed with others that also represent the same `exog` variable, but for a different series.

![imagen](https://github.com/JoaquinAmatRodrigo/skforecast/assets/95030312/2d13eb8a-40d8-42e3-aac7-d9d339fad6e8)




We have this feature in the backlog, but the only use case that comes to my mind that needs this is for example modeling a time series by countries and you need a different vacation indicator for each of them. Could you give us some information about your use case? 

Thank you very much!

Javier

**Reply by yarnabrina** [reply]
Hi @JavierEscobarOrtiz ,

Thanks for taking a look. My main problem regarding (3) is not that I have different exogenous variables per series. I have same variables, but they also vary based on the series itself. So continuing the example from your second picture, I'll expect `a-g` to vary from series 1 and series 2, i.e. `series 1` will have `[a, b, c, d, e, f, g]` and `series 2` will have `[h, i, j, k, l, m, n]`.

My particular use case which I was trying `skforecast` for deals with financial forecasting, but I'm currently not at liberty to share any more details of that. But let's consider an example use case of demand forecasting problem of different products in a store. Sales is the endogenous variable, and retail price (subject to campaigns/promotions etc.) is the exogenous variable, and of course both sales/price is varying per every product (or per category, or per store/product combination, etc.). I'll actually consider that different exogenous variables (values, not the variables itselves, but even that can be a generalisation) per different series is more common.

I could not find a way to pass in this way (different exogenous values per series) to `ForecasterAutoregMultiSeries`, so I guess as of now I've to use indivual `ForecasterAutoreg` per series. Can you please confirm?

**Comment by JoaquinAmatRodrigo** [comment]
Hi @yarnabrina,
Thank you for your detailed explanation. Unfortunately, the current implementation of `ForecasterAutoregMultiSeries` does not allow different exogenous variables (or the same exogenous but different values) per series. This is a feature already requested by several skforecast users, so we will work on it as soon as possible. Feel free to suggest implementations or ideas.

**Reply by yarnabrina** [reply]
Thanks for the confirmation @JoaquinAmatRodrigo, for now let me restrict to `ForecasterAutoreg` only.

I'm not familiar with the `skforecast` codebase, so don't think I can suggest anything with respect to implementation. But based on the 2nd image @JavierEscobarOrtiz shared above, why not just use the same format as in middle and right one, but replace the values for series 2 from a-g to h-n?

In my case of same exogenous features but different values per series, I've dataset already in the long format (series_id, time_id, target_var, exog_var_1, exog_var_2, etc.). This is same as the middle one in above, and converting from that to right one seems to be straighforward. Your existing code to create dummies for series identifier should be able to create that hopefully without any changes.

**Comment by JoaquinAmatRodrigo** [comment]
We seem to agree in our understanding of the situation and are considering starting soon with the scenario you describe (**same exogenous features but different values per series**). Once we have developed a first version, perhaps you can help us evaluate its effectiveness.

**Comment by JavierEscobarOrtiz** [comment, marked as answer]
This feature was added as part of skforecast 0.13.0. Closing the discussion

https://skforecast.org/latest/user_guides/multi-series-with-different-length-and-different_exog
